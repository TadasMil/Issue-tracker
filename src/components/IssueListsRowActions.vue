<script setup lang="ts" name="IssueListsRowActions">
import { useAttrs } from 'vue'

const attrs = useAttrs()
</script>

<template>
  <b-dropdown
    id="dropdown"
    size="lg"
    variant="link"
    aria-labelledby="dropdown"
    toggle-class="text-decoration-none"
    dropleft
    no-caret
  >
    <template #button-content>
      <div class="text-dark">
        <svg width="32" height="32" viewBox="0 0 32 32">
          <path
            fill="none"
            stroke="currentColor"
            stroke-width="2"
            d="M16 24a1 1 0 1 0 0-2a1 1 0 0 0 0 2Zm0-7a1 1 0 1 0 0-2a1 1 0 0 0 0 2Zm0-7a1 1 0 1 0 0-2a1 1 0 0 0 0 2Z"
          />
        </svg>
      </div>
    </template>

    <slot />

    <b-dropdown-item v-if="attrs.onEdit" @click="attrs.onEdit">
      Edit
      <i class="bi bi-pencil fa-1x text-success"></i>
    </b-dropdown-item>

    <b-dropdown-item
      v-if="attrs.onMarkAsTrashed"
      @click="attrs.onMarkAsTrashed"
    >
      Mark as trashed
      <i class="bi bi-trash fa-1x text-danger"></i>
    </b-dropdown-item>

    <b-dropdown-item v-if="attrs.onMarkAsOpen" @click="attrs.onMarkAsOpen">
      Mark as open
      <i class="bi bi-repeat fa-1x text-success"></i>
    </b-dropdown-item>

    <b-dropdown-item
      v-if="attrs.onMarkAsCompleted"
      @click="attrs.onMarkAsCompleted"
    >
      Mark as completed
      <i class="bi bi-check-circle fa-1x text-success"></i>
    </b-dropdown-item>

    <b-dropdown-item
      v-if="attrs.onRestoreTrashed"
      @click="attrs.onRestoreTrashed"
    >
      Restore trashed issue
      <i class="bi bi-trash fa-1x text-danger"></i>
    </b-dropdown-item>
  </b-dropdown>
</template>
