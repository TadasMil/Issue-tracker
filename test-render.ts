import type { RenderOptions } from '@testing-library/vue'
import BootstrapVue3 from 'bootstrap-vue-3'
import { render } from '@testing-library/vue'

const customRender = (ui: any, options?: RenderOptions) =>
  render(ui, {
    ...options,
    global: {
      ...options?.global,
      plugins: [BootstrapVue3, ...(options?.global?.plugins || [])],
    },
  })

export * from '@testing-library/vue'
export { customRender as render }
