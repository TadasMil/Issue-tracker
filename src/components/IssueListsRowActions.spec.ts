import { describe, it, expect, vi } from 'vitest'
import { render, fireEvent } from '../../test-render'
import '@testing-library/jest-dom'

import IssueListsRowActions from './IssueListsRowActions.vue'

const cases = [
  {
    name: 'Edit',
    event: 'onEdit',
  },
  {
    name: 'Mark as trashed',
    event: 'onMarkAsTrashed',
  },
  {
    name: 'Mark as open',
    event: 'onMarkAsOpen',
  },
  {
    name: 'Mark as completed',
    event: 'onMarkAsCompleted',
  },
  {
    name: 'Restore trashed issue',
    event: 'onRestoreTrashed',
  },
]

describe('IssueListsRowActions', () => {
  it('should render the component default component', () => {
    const { html } = render(IssueListsRowActions)

    expect(html()).toMatchSnapshot()
  })

  cases.forEach((label) => {
    it(`should not render ${label} button`, () => {
      const { queryByText } = render(IssueListsRowActions)

      expect(queryByText(label.name)).not.toBeInTheDocument()
    })
  })

  cases.forEach((c) => {
    it(`should call ${c.event} when ${c.name} is clicked`, () => {
      const eventFn = vi.fn()

      const { getByText, html } = render(IssueListsRowActions, {
        attrs: {
          [c.event]: eventFn,
        },
      })

      fireEvent.click(getByText(c.name))

      expect(eventFn).toHaveBeenCalled()
      expect(html()).toMatchSnapshot()
    })
  })
})
