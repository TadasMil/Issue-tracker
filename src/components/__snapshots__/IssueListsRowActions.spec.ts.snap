// Vitest Snapshot v1

exports[`IssueListsRowActions > should call onEdit when Edit is clicked 1`] = `
"<div class=\\"btn-group\\" aria-labelledby=\\"dropdown\\"><button class=\\"btn btn-link btn-lg text-decoration-none dropdown-toggle dropdown-toggle-no-caret\\" type=\\"button\\" id=\\"dropdown\\" data-bs-toggle=\\"dropdown\\" aria-expanded=\\"false\\">
    <!---->
    <div class=\\"btn-content\\">
      <div class=\\"text-dark\\"><svg width=\\"32\\" height=\\"32\\" viewBox=\\"0 0 32 32\\">
          <path fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" d=\\"M16 24a1 1 0 1 0 0-2a1 1 0 0 0 0 2Zm0-7a1 1 0 1 0 0-2a1 1 0 0 0 0 2Zm0-7a1 1 0 1 0 0-2a1 1 0 0 0 0 2Z\\"></path>
        </svg></div>
    </div>
  </button>
  <!---->
  <ul class=\\"dropdown-menu\\" aria-labelledby=\\"dropdown\\" role=\\"menu\\">
    <li role=\\"presentation\\" class=\\"\\"><button class=\\"dropdown-item\\" type=\\"button\\" target=\\"_self\\"> Edit <i class=\\"bi bi-pencil fa-1x text-success\\"></i></button></li>
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
  </ul>
</div>"
`;

exports[`IssueListsRowActions > should call onMarkAsCompleted when Mark as completed is clicked 1`] = `
"<div class=\\"btn-group\\" aria-labelledby=\\"dropdown\\"><button class=\\"btn btn-link btn-lg text-decoration-none dropdown-toggle dropdown-toggle-no-caret\\" type=\\"button\\" id=\\"dropdown\\" data-bs-toggle=\\"dropdown\\" aria-expanded=\\"false\\">
    <!---->
    <div class=\\"btn-content\\">
      <div class=\\"text-dark\\"><svg width=\\"32\\" height=\\"32\\" viewBox=\\"0 0 32 32\\">
          <path fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" d=\\"M16 24a1 1 0 1 0 0-2a1 1 0 0 0 0 2Zm0-7a1 1 0 1 0 0-2a1 1 0 0 0 0 2Zm0-7a1 1 0 1 0 0-2a1 1 0 0 0 0 2Z\\"></path>
        </svg></div>
    </div>
  </button>
  <!---->
  <ul class=\\"dropdown-menu\\" aria-labelledby=\\"dropdown\\" role=\\"menu\\">
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
    <li role=\\"presentation\\" class=\\"\\"><button class=\\"dropdown-item\\" type=\\"button\\" target=\\"_self\\"> Mark as completed <i class=\\"bi bi-check-circle fa-1x text-success\\"></i></button></li>
    <!--v-if-->
  </ul>
</div>"
`;

exports[`IssueListsRowActions > should call onMarkAsOpen when Mark as open is clicked 1`] = `
"<div class=\\"btn-group\\" aria-labelledby=\\"dropdown\\"><button class=\\"btn btn-link btn-lg text-decoration-none dropdown-toggle dropdown-toggle-no-caret\\" type=\\"button\\" id=\\"dropdown\\" data-bs-toggle=\\"dropdown\\" aria-expanded=\\"false\\">
    <!---->
    <div class=\\"btn-content\\">
      <div class=\\"text-dark\\"><svg width=\\"32\\" height=\\"32\\" viewBox=\\"0 0 32 32\\">
          <path fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" d=\\"M16 24a1 1 0 1 0 0-2a1 1 0 0 0 0 2Zm0-7a1 1 0 1 0 0-2a1 1 0 0 0 0 2Zm0-7a1 1 0 1 0 0-2a1 1 0 0 0 0 2Z\\"></path>
        </svg></div>
    </div>
  </button>
  <!---->
  <ul class=\\"dropdown-menu\\" aria-labelledby=\\"dropdown\\" role=\\"menu\\">
    <!--v-if-->
    <!--v-if-->
    <li role=\\"presentation\\" class=\\"\\"><button class=\\"dropdown-item\\" type=\\"button\\" target=\\"_self\\"> Mark as open <i class=\\"bi bi-repeat fa-1x text-success\\"></i></button></li>
    <!--v-if-->
    <!--v-if-->
  </ul>
</div>"
`;

exports[`IssueListsRowActions > should call onMarkAsTrashed when Mark as trashed is clicked 1`] = `
"<div class=\\"btn-group\\" aria-labelledby=\\"dropdown\\"><button class=\\"btn btn-link btn-lg text-decoration-none dropdown-toggle dropdown-toggle-no-caret\\" type=\\"button\\" id=\\"dropdown\\" data-bs-toggle=\\"dropdown\\" aria-expanded=\\"false\\">
    <!---->
    <div class=\\"btn-content\\">
      <div class=\\"text-dark\\"><svg width=\\"32\\" height=\\"32\\" viewBox=\\"0 0 32 32\\">
          <path fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" d=\\"M16 24a1 1 0 1 0 0-2a1 1 0 0 0 0 2Zm0-7a1 1 0 1 0 0-2a1 1 0 0 0 0 2Zm0-7a1 1 0 1 0 0-2a1 1 0 0 0 0 2Z\\"></path>
        </svg></div>
    </div>
  </button>
  <!---->
  <ul class=\\"dropdown-menu\\" aria-labelledby=\\"dropdown\\" role=\\"menu\\">
    <!--v-if-->
    <li role=\\"presentation\\" class=\\"\\"><button class=\\"dropdown-item\\" type=\\"button\\" target=\\"_self\\"> Mark as trashed <i class=\\"bi bi-trash fa-1x text-danger\\"></i></button></li>
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
  </ul>
</div>"
`;

exports[`IssueListsRowActions > should call onRestoreTrashed when Restore trashed issue is clicked 1`] = `
"<div class=\\"btn-group\\" aria-labelledby=\\"dropdown\\"><button class=\\"btn btn-link btn-lg text-decoration-none dropdown-toggle dropdown-toggle-no-caret\\" type=\\"button\\" id=\\"dropdown\\" data-bs-toggle=\\"dropdown\\" aria-expanded=\\"false\\">
    <!---->
    <div class=\\"btn-content\\">
      <div class=\\"text-dark\\"><svg width=\\"32\\" height=\\"32\\" viewBox=\\"0 0 32 32\\">
          <path fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" d=\\"M16 24a1 1 0 1 0 0-2a1 1 0 0 0 0 2Zm0-7a1 1 0 1 0 0-2a1 1 0 0 0 0 2Zm0-7a1 1 0 1 0 0-2a1 1 0 0 0 0 2Z\\"></path>
        </svg></div>
    </div>
  </button>
  <!---->
  <ul class=\\"dropdown-menu\\" aria-labelledby=\\"dropdown\\" role=\\"menu\\">
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
    <li role=\\"presentation\\" class=\\"\\"><button class=\\"dropdown-item\\" type=\\"button\\" target=\\"_self\\"> Restore trashed issue <i class=\\"bi bi-trash fa-1x text-danger\\"></i></button></li>
  </ul>
</div>"
`;

exports[`IssueListsRowActions > should render the component default component 1`] = `
"<div class=\\"btn-group\\" aria-labelledby=\\"dropdown\\"><button class=\\"btn btn-link btn-lg text-decoration-none dropdown-toggle dropdown-toggle-no-caret\\" type=\\"button\\" id=\\"dropdown\\" data-bs-toggle=\\"dropdown\\" aria-expanded=\\"false\\">
    <!---->
    <div class=\\"btn-content\\">
      <div class=\\"text-dark\\"><svg width=\\"32\\" height=\\"32\\" viewBox=\\"0 0 32 32\\">
          <path fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" d=\\"M16 24a1 1 0 1 0 0-2a1 1 0 0 0 0 2Zm0-7a1 1 0 1 0 0-2a1 1 0 0 0 0 2Zm0-7a1 1 0 1 0 0-2a1 1 0 0 0 0 2Z\\"></path>
        </svg></div>
    </div>
  </button>
  <!---->
  <ul class=\\"dropdown-menu\\" aria-labelledby=\\"dropdown\\" role=\\"menu\\"></ul>
</div>"
`;
